command/head.go

```diff
@@ -0,0 +1,313 @@
+				Aliases: []string{"H"},
+				Usage:   "human-readable output for object sizes",
+			},
+			&cli.BoolFlag{
+				Name:    "storage-class",
+				Aliases: []string{"s"},
+				Usage:   "display full name of the object class",
+				Value:   true,
+			},
+			&cli.BoolFlag{
+				Name:  "show-fullpath",
+				Usage: "shows only the fullpath names of the object(s)",
+			},
+			&cli.StringFlag{
+				Name:  "version-id",
+				Usage: "use the specified version of an object",
+			},
+			&cli.BoolFlag{
+				Name:  "raw",
+				Usage: "disable the wildcard operations, useful with filenames that contains glob characters",
+			},
+		},
+
+		Before: func(c *cli.Context) error {
+			err := validateHeadCommand(c)
+			if err != nil {
+				printError(commandFromContext(c), c.Command.Name, err)
+			}
+			return err
+		},
+		Action: func(c *cli.Context) (err error) {
+			defer stat.Collect(c.Command.FullName(), &err)()
+
+			op := c.Command.Name
+			fullCommand := commandFromContext(c)
+			src, err := url.New(c.Args().Get(0), url.WithVersion(c.String("version-id")),
+				url.WithRaw(c.Bool("raw")))
+			if err != nil {
+				printError(fullCommand, op, err)
+				return err
+			}
+
+			return Head{
+				src:         src,
+				op:          op,
+				fullCommand: fullCommand,
+				// flags
+				showEtag:         c.Bool("etag"),
+				humanize:         c.Bool("humanize"),
+				showStorageClass: c.Bool("storage-class"),
+				showFullPath:     c.Bool("show-fullpath"),
+
+				storageOpts: NewStorageOpts(c),
+			}.Run(c.Context)
+		},
+	}
+	cmd.BashComplete = getBashCompleteFn(cmd, true, false)
+	return cmd
+}
+
+type Head struct {
+	src         *url.URL
+	op          string
+	fullCommand string
+
+	showEtag         bool
+	humanize         bool
+	showStorageClass bool
+	showFullPath     bool
+
+	storageOpts storage.Options
+}
+
+func (h Head) Run(ctx context.Context) error {
+	h.src.SetRelative(h.src)
+	client, err := storage.NewRemoteClient(ctx, h.src, h.storageOpts)
+	if err != nil {
+		printError(h.fullCommand, h.op, err)
+		return err
+	}
+
+	if h.src.IsBucket() {
+		err := client.HeadBucket(ctx, h.src)
+		if err != nil {
+			printError(h.fullCommand, h.op, err)
+			return err
+		}
+
+		msg := HeadBucketMessage{
+			Name: h.src.String(),
+		}
+
+		log.Info(msg)
+
+		return nil
+	}
+
+	object, metadata, err := client.HeadObject(ctx, h.src)
+	if err != nil {
+		printError(h.fullCommand, h.op, err)
+		return err
+	}
+
+	msg := HeadObjectMessage{
+		Object:           object,
+		showEtag:         h.showEtag,
+		showHumanized:    h.humanize,
+		showStorageClass: h.showStorageClass,
+		showFullPath:     h.showFullPath,
+		Metadata:         metadata,
+	}
+
+	log.Info(msg)
+
+	return nil
+}
+
+type HeadObjectMessage struct {
+	Object *storage.Object `json:"object"`
+
+	showEtag         bool
+	showHumanized    bool
+	showStorageClass bool
+	showFullPath     bool
+	Metadata         map[string]string `json:"metadata"`
+}
+
+func (m HeadObjectMessage) String() string {
```

Comment:
Since the head command is used for a single object or a single bucket, let’s format the output as json and print all the data received from S3. The columnar format we discussed earlier doesn’t make much sense for a single row. Sorry for frequent changes in the directions  `¯\_(ツ)_/¯`

Reply:
SGTM.